docs/design/api-guidelines/api-guidelines.md

```diff
@@ -0,0 +1,233 @@
+
+The document is addressed to developers who design or enhance Northbound API.
+
+# Principles
+
+### OpenKilda REST API Design Principles
+
+- Simplicity, comprehensibility, and usability of APIs.
+- Prefer REST-based APIs with JSON payload.
+- Follow the RESTful architectural style as much as possible. 
+
+    Although per definition RESTful API has to support HATEOAS (maturity level 3). Our guidelines don't enforce full RESTful compliance, but limited hypermedia support. However, we use "RESTful API" to refer implementation of the [REST Maturity level 2](https://martinfowler.com/articles/richardsonMaturityModel.html#level2).
+
+### What is RESTful API interface?
+
+The RESTful architectural style describes six constraints:
+* Client-Server    
+* Stateless
+* Cacheable/switches/{switch_id}/port
+* Layered System
+* Code on Demand (optional)
+* Uniform Interface
+
+A web service which violates any of the required constraints **cannot be considered RESTful**.
+
+For a more detailed overview, check out:
+- [REST on Wikipedia](https://en.wikipedia.org/wiki/Representational_state_transfer)
+- [Architectural Styles and the Design of Network-based Software Architectures](https://www.ics.uci.edu/~fielding/pubs/dissertation/rest_arch_style.htm) - the chapter on REST in Roy Fielding's dissertation on Network Architecture.
+- [RFC 7231](https://tools.ietf.org/html/rfc7231)
+
+# Conventions & Rules
+
+### Resource
+
+URIs follow [RFC 3986](https://tools.ietf.org/html/rfc3986) specification.
+
+#### URI Naming Conventions
+
+[Resource Naming](https://www.restapitutorial.com/lessons/restfulresourcenaming.html)
+
+* Use nouns and NOT the verbs.
+* Use the plural version of a resource name to be consistent when referring to particular resources, e.g.:
+```
+/switches
+/flows
+/links
+```
+* Use lowercase and dash-separated path names, e.g.:
+```
+/health-check
+```
+* Use snake_case (underscore-separated) query string literals.
+
+### Use HTTP methods to operate on collections and entities
+
+[HTTP methods](https://www.restapitutorial.com/lessons/httpmethods.html)
+
+E.g.
+
+| Resource / HTTP method | POST (create)    | GET (read)  | PUT (update)           | DELETE (delete)    |
+| ---------------------- | ---------------- | ----------- | ---------------------- | ------------------ |
+| /flows                 | Create new flow  | List flows  | Error                  | Error              |
+| /flows/{flow_id}       | Error            | Get flow    | Update flow if exists  | Delete flow        |
+
+
+Update and create must return a resource representation.
+
+#### Create entity
+
+* Server returns the `HTTP 201 Created` status code
+* Response will also include the `Location` header that points to the URL of the new resource ([RFC 6570](http://tools.ietf.org/html/rfc6570))
+* Server returns the new entity
+
+#### List entities (collection)
+
+* Response collection is presented as a plain array, not a dictionary with the array as a value.
```

Comment:
You should choose something one:
```
#### Wrap collection in object
 Always return root element as an object. This way you can add extra fields to the response without compatibility breakdown.
```

From my point of view - you must return `object` instead of `list`. At least it will be much easier to add pagination, when you will need to.

Reply:
Nice catch! These two for sure contradict each other. I'll fix it.

Reply:
Fixed.